overlord/devicestate/handlers_gadget.go

```diff
@@ -125,13 +144,35 @@ func (m *DeviceManager) doUpdateGadgetAssets(t *state.Task, _ *tomb.Tomb) error
 		return nil
 	}
 
+	// add kernel directories
+	currentKernelInfo, err := snapstate.CurrentInfo(st, groundDeviceCtx.Model().Kernel())
+	// XXX: switch to the normal `if err != nil { return err }` pattern
+	// here once all tests are updated and have a kernel
+	if err == nil {
+		currentData.KernelRootDir = currentKernelInfo.MountDir()
+		updateData.KernelRootDir = currentKernelInfo.MountDir()
+	}
+	// if this is a gadget update triggered by an updated kernel we
+	// need to ensure "updateData.KernelRootDir" points to the new kernel
+	if snapsup.Type == snap.TypeKernel {
+		updateKernelInfo, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+		if err != nil {
+			return fmt.Errorf("cannot read candidate kernel snap details: %v", err)
+		}
+		updateData.KernelRootDir = updateKernelInfo.MountDir()
+	}
+
 	snapRollbackDir, err := makeRollbackDir(fmt.Sprintf("%v_%v", snapsup.InstanceName(), snapsup.SideInfo.Revision))
 	if err != nil {
 		return fmt.Errorf("cannot prepare update rollback directory: %v", err)
 	}
 
 	var updatePolicy gadget.UpdatePolicyFunc = nil
 
+	if snapsup.Type == snap.TypeKernel {
+		updatePolicy = gadget.KernelUpdatePolicy
```

Comment:
what should happen if it's both a kernel update and a remodel?

Reply:
A remodel should be fine, I guess this needs a test but the update policy always returns true so everything will be updated here.

Reply:
-I think the question is whether it makes sense for the kernel part of the remodel to touch the gadget-only assets? I'm not sure that updating the gadget-only assets from the kernel task makes total sense.

OTOH: in theory  if we have new gadget or  an edition bump of the gadget,  we would like to update kernel, kernel assets, gadget assets all together. That is not something that we can model well right now, nor do well.

Reply:
Indeed, I think we can for now just make sure that a kernel asset update always uses the gadget.KernelUpdatePolicy even when a remodel is in progress. It feels much cleaner this way, i.e. kernel always takes care of kernel bits only.